### ISSUE TITLE
how to close stdin 

### ISSUE THREAD

**Issue by dcolens** [apocas/dockerode#136]
I'm trying to create a container, launch a process (let's say 'wc -c' for the example), send data to its stdin and read the response (number of bytes in this example).

It all works fine if I use:
```
      optsc.AttachStdin = true;
      optsc.AttachStout = true;
      optsc.AttachSterr = true;
      optsc.Tty = true;
      optsc.OpenStdin = true;
```
as long as the data written to stdin is less than 4096 bytes, after that it goes wrong, i.e. wc returns "1" when I send 4097 bytes. I also tried writing by smaller chunks using a bufferstream to respect the stream's readiness.

I tried not using tty=true, but in that case I can't figure how to send EOT to stdin and the command never exits. I tried using writing '\n\x04', using stream.end or stream.destroy as per #53 without success.

I guess I'm not doing this right, but not too sure how, the example uses the tty=true, the unit test does not use EOT (it uses bash's exit command).

sample code (based on the mocha unit tests) to reproduce the issue :
```
    it("should support attach with stdin enable", function(done) {
      this.timeout(120000);

      var length = 4096; //set to 4095 to see it working fine, note that stdin will be 1char longer because of the terminating `\nctrl+d`

      function randomString(length) {
        var result = '', chars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
        for (var i = length; i > 0; --i) result += chars[Math.round(Math.random() * (chars.length - 1))];
        return result;
      }

      function handler(err, container) {
        expect(err).to.be.null;
        expect(container).to.be.ok;

        var attach_opts = {
          stream: true,
          stdin: true,
          stdout: true,
          stderr: true
        };
        container.attach(attach_opts, function handler(err, stream) {
          expect(err).to.be.null;
          expect(stream).to.be.ok;

          var memStream = new MemoryStream();
          var output = '';
          memStream.on('data', function(data) {
            output += data.toString();
          });
          memStream.on('end', function(data) {
            console.log('memStream.end');
          });
          stream.pipe(memStream);

          
          
          container.start(function(err, data) {
            expect(err).to.be.null;
            stream.write(randomString(length)+'\n\x04');

            container.wait(function(err, data) {
              expect(err).to.be.null;
              expect(data).to.be.ok;
              output = output.trim();
              console.log('received:'+output);
              expect(output).to.match(new RegExp('.*'+(length+1)+'$'));
              done();
            });
          });      
        });
      }

      optsc.AttachStdin = true;
      optsc.AttachStout = true;
      optsc.AttachSterr = true;
      optsc.Tty = true
      optsc.OpenStdin = true;

      optsc.Cmd = ['wc', '-c'];

      docker.createContainer(optsc, handler);
    });
```

Any sample code that would allow me to run 'wc -c', sending more than 4096 bytes to stdin and reading the number of bytes from stdout would be great :)

Did

**Comment by apocas** [comment]
Check this out https://github.com/apocas/dockerode/issues/53

**Comment by dcolens** [comment]
Thanks,

for the quick response, I knew about #53 and tried stream.end and stream.destroy without success (docker-modem@0.2.2). I guess I don’t understand #53, could you help me understand how I should use the destroy method ?

Thanks,

Did


>

**Comment by dcolens** [comment]
Any other advice? with tty=true it fails after 4096 bytes sent via stdin and with tty=false it never receives stdin. what's the proper way to send EOT with tty=false?

**Comment by apocas** [comment]
Verify that you really have an HttpDuplex object and the end method should work.

No need to use destroy.

**Comment by dcolens** [comment]
I disagree with you this is just a question, as far as I can tell these are 2 bugs. It is trivial to repro. Just try running the following code :

'use strict';
var util = require('util');
var assert = require('assert');
var Docker = require('dockerode');
var docker = new Docker();
var MemoryStream = require('memorystream');
var stdinSize = 4097;
var optsc = {
  'Hostname': '',
  'User': '',
  'AttachStdin': true,
  'AttachStdout': true,
  'AttachStderr': true,
  'Tty': false,
  'OpenStdin': true,
  'StdinOnce': false,
  'Env': null,
  'Cmd': ['wc', '-c'],
  'Dns': ['8.8.8.8', '8.8.4.4'],
  'Image': 'ubuntu',
  'Volumes': {},
  'VolumesFrom': ''
};

function randomString(size) {
  var result = '', chars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
  for (var i = size; i > 0; --i) { result += chars[Math.round(Math.random() * (chars.length - 1))]; }
  return result;
}

function handler(err, container) {

  var attach_opts = {
    stream: true,
    stdin: true,
    stdout: true,
    stderr: true
  };
  container.attach(attach_opts, function handler(err, stream) {

    var memStream = new MemoryStream();
    var output = '';
    memStream.on('data', function(data) {
      output += data.toString();
    });
    memStream.on('end', function() {
      console.log('memStream.end, received: '+output.length+' bytes');
    });

    stream.pipe(memStream);

    container.start(function() {

      stream.end(randomString(stdinSize));
      container.wait(function() {
        output = output.trim();
        console.log('received:'+output);
        assert(new RegExp('.*'+(stdinSize)+'$').test(output));
      });
    });
  });
}

docker.createContainer(optsc, handler);


I am sure that the above stream is an instance of httpduplex. It would be great to get a useful suggestion.

Did

**Comment by apocas** [comment]
I'm using end in production without any problems since it was implemented in https://github.com/apocas/dockerode/issues/53, although can't confirm Docker versions or anything right now.

When I have time I will try to reproduce your example. Don't know when it will be.

If you are in a hurry try to do some debug in https://github.com/apocas/docker-modem/blob/master/lib/http_duplex.js to get some details on the issue, since stdin is managed in here.

**Comment by apocas** [comment]
Confirmed by https://github.com/apocas/dockerode/pull/146

**Comment by apocas** [comment]
Check the tests now.

The was not being truncated it is sent into the container, was just being sent in multiple "chunks" into "wc" which mess things up.
I rewrote the test check it out, it even has the tty disabled like you wanted so it also answers your second doubt.

This use case is a bit... "tricky", dockerode is the simplest part in all this it just send the data. You have the data being sent, piped, processed, displayed...

I was seeing in the stdout stream the 5k bytes and when dumping them into a file I saw them inside the container filesystem, so I just rewrote your docker usage keeping the same philosophy of sending the data in the duplexstream :)
